### PR TITLE
rafthttp: improve error printout

### PR DESCRIPTION
**Pull request by yichengq**
1. use uniform log format
2. improve log in stream
3. not print duplicate log line

The log to start 2 members in 3-member cluster:
```
13:30:47 etcd1 | 2015/02/25 13:30:47 rafthttp: encountered error in stream msgapp from a8266ecf031671f3 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:30:47 etcd2 | 2015/02/25 13:30:47 rafthttp: encountered error in stream msgapp from 6e3bd23ae5f1eae0 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:30:47 etcd2 | 2015/02/25 13:30:47 rafthttp: encountered error in stream message from 6e3bd23ae5f1eae0 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:30:49 etcd2 | 2015/02/25 13:30:49 rafthttp: encountered error in pipeline from 6e3bd23ae5f1eae0 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:30:57 etcd1 | 2015/02/25 13:30:57 rafthttp: encountered error in stream msgapp from a8266ecf031671f3 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:30:57 etcd1 | 2015/02/25 13:30:57 rafthttp: encountered error in stream message from a8266ecf031671f3 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:30:57 etcd2 | 2015/02/25 13:30:57 rafthttp: encountered error in stream message from 6e3bd23ae5f1eae0 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:30:57 etcd2 | 2015/02/25 13:30:57 rafthttp: encountered error in stream msgapp from 6e3bd23ae5f1eae0 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:30:59 etcd2 | 2015/02/25 13:30:59 rafthttp: encountered error in pipeline from 6e3bd23ae5f1eae0 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:31:07 etcd1 | 2015/02/25 13:31:07 rafthttp: encountered error in stream message from a8266ecf031671f3 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:31:07 etcd1 | 2015/02/25 13:31:07 rafthttp: encountered error in stream msgapp from a8266ecf031671f3 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:31:08 etcd2 | 2015/02/25 13:31:08 rafthttp: encountered error in stream msgapp from 6e3bd23ae5f1eae0 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:31:08 etcd2 | 2015/02/25 13:31:08 rafthttp: encountered error in stream message from 6e3bd23ae5f1eae0 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:31:09 etcd2 | 2015/02/25 13:31:09 rafthttp: encountered error in pipeline from 6e3bd23ae5f1eae0 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:31:17 etcd1 | 2015/02/25 13:31:17 rafthttp: encountered error in stream msgapp from a8266ecf031671f3 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:31:17 etcd1 | 2015/02/25 13:31:17 rafthttp: encountered error in stream message from a8266ecf031671f3 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:31:18 etcd2 | 2015/02/25 13:31:18 rafthttp: encountered error in stream message from 6e3bd23ae5f1eae0 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:31:18 etcd2 | 2015/02/25 13:31:18 rafthttp: encountered error in stream msgapp from 6e3bd23ae5f1eae0 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:31:19 etcd2 | 2015/02/25 13:31:19 rafthttp: encountered error in pipeline from 6e3bd23ae5f1eae0 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
```

### CODE REVIEW

File: rafthttp/peer.go [hunk 1]
```diff
@@ -33,6 +34,17 @@ const (
 	recvBufSize = 4096
 )
 
+type status struct {
+	name   string
+	active bool
+	*opError
+}
+
+type worker struct {
```
Comment (xiang90) [inline rafthttp/peer.go]: It is hard to know what the work exactly is. Can you use a more specific name for the "worker"?

File: rafthttp/peer.go [hunk 2]
```diff
@@ -192,3 +239,46 @@ func (p *peer) pick(m raftpb.Message) (writec chan raftpb.Message, name string,
 	}
 	return
 }
+
+// opError is the error type passed in the package that describing the
+// operation, worker type, remote id and address of an error.
+type opError struct {
```
Comment (xiang90) [inline rafthttp/peer.go]: move to op_error.go

File: rafthttp/peer.go [hunk 3]
```diff
@@ -100,6 +117,36 @@ func startPeer(tr http.RoundTripper, u string, local, to, cid types.ID, r Raft,
 				if err := r.Process(context.TODO(), mm); err != nil {
 					log.Printf("peer: process raft message error: %v", err)
 				}
+			case status := <-p.statusc:
+				name := status.name
```
Comment (xiang90) [inline rafthttp/peer.go]: these all "work" related logic, should not be put into the peer code. should have a separate function to abstract out the logic.

**Comment by yichengq** [comment]
PTAL

File: rafthttp/peer.go [hunk 4]
```diff
@@ -65,9 +65,9 @@ type peer struct {
 func startPeer(tr http.RoundTripper, u string, local, to, cid types.ID, r Raft, fs *stats.FollowerStats, errorc chan error) *peer {
 	p := &peer{
 		id:           to,
-		msgAppWriter: startStreamWriter(fs),
-		writer:       startStreamWriter(fs),
-		pipeline:     newPipeline(tr, u, to, cid, fs, errorc),
+		msgAppWriter: startStreamWriter(local, to, streamTypeMsgApp, fs),
```
Comment (xiang90) [inline rafthttp/peer.go]: local, remote
or from, to

not local, to

File: rafthttp/status.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	statusMapMu sync.Mutex
+	statusMap   = make(map[string]connStatus)
+)
+
+type connStatus struct {
+	active bool
+	errm   map[string]bool
+}
+
+func report(local, to types.ID, src string, active bool, err *opError) {
```
Comment (xiang90) [inline rafthttp/status.go]: reportLinkStatus

File: rafthttp/status.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
+// Copyright 2015 CoreOS, Inc.
```
Comment (xiang90) [inline rafthttp/status.go]: linkstatus.go

File: rafthttp/status.go (hunk 5 above)
Comment (xiang90) [inline rafthttp/status.go]: or connStatus

Comment (xiang90) [inline rafthttp/status.go]: can we change the bool to (connStatusOK, connStatusBad)?

Comment (yichengq) [inline rafthttp/status.go]: it truly can be confusing to put the bool parameter here. will do

File: rafthttp/status.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+type connStatus struct {
+	active bool
+	errm   map[string]bool
+}
+
+func report(local, to types.ID, src string, active bool, err *opError) {
+	statusMapMu.Lock()
+	defer statusMapMu.Unlock()
+	name := fmt.Sprintf("%s %s %s", local, to, src)
```
Comment (xiang90) [inline rafthttp/status.go]: this is terrible. the key of a map can be a struct. do not print it out as the key

**Comment by yichengq** [comment]
ready for the next review

**Comment by yichengq** [comment]
ready for the next round.

File: rafthttp/linkstatus.go [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+	statusOK  status = true
+	statusBad status = false
+)
+
+type linkState struct {
+	status status
+	errm   map[string]bool
+}
+
+func reportLinkStatus(local, remote types.ID, src string, status status, err *opError) {
```
Comment (xiang90) [inline rafthttp/linkstatus.go]: why it is called "src"?

Comment (xiang90) [inline rafthttp/linkstatus.go]: do we need to pass in an opError type?

Comment (xiang90) [inline rafthttp/linkstatus.go]: moreover, if we already pass in the error why do we need the status field? error = nil implies statusOK?

File: rafthttp/linkstatus.go [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+		if strings.HasPrefix(src, "stream") {
+			if status == statusOK {
+				log.Printf("rafthttp: %s from %s to %s becomes active", src, local, remote)
+			} else {
+				log.Printf("rafthttp: %s from %s to %s becomes inactive", src, local, remote)
+			}
+		}
+		linkStateMap[name] = linkState{status: status, errm: make(map[string]bool)}
+	}
+	// TODO: print out the error still after some timeout
```
Comment (xiang90) [inline rafthttp/linkstatus.go]: TODO: do not ignore the error after given timeout

File: rafthttp/linkstatus.go [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+type status bool
+
+const (
+	statusOK  status = true
+	statusBad status = false
+)
+
+type linkState struct {
+	status status
+	errm   map[string]bool
```
Comment (xiang90) [inline rafthttp/linkstatus.go]: why do we need a map here? we probably only want to record the last error.

File: rafthttp/linkstatus.go [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+type status bool
+
+const (
+	statusOK  status = true
+	statusBad status = false
+)
+
+type linkState struct {
+	status status
```
Comment (xiang90) [inline rafthttp/linkstatus.go]: we can get rid of the status.

File: rafthttp/report.go [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+)
+
+type linkName struct {
+	local, remote types.ID
+	worker        string
+}
+
+type linkErr struct {
+	err        *opError
+	recordTime time.Time
```
Comment (xiang90) [inline rafthttp/report.go]: remove the time suffix?

Comment (xiang90) [inline rafthttp/report.go]: last time.Time?

**Comment by yichengq** [comment]
PTAL

**Comment by barakmich** [comment]
So I'm all for a good refactor, but if it's actually an error, and it keeps happening, why are we suppressing it? Because it's either (a) normal-ish operation (failures predicted and well-handled) for which this is a higher log level, or (b) actually an error!

**Comment by xiang90** [comment]
@barakmich raft and our current transport connection is kind of passive. For lower recovery latency, one side will keep on probing for the other failed side. We do not want to log every error the probing generated.

**Comment by yichengq** [comment]
It is an actual error.
We suppress it because we don't want the error is flooded out. Or it will print out the full screen in one second and is not helpful.

**Comment by barakmich** [comment]
I get what you're saying, but I think part of the thing I'm saying is that we don't have log levels. One would expect every error at the Debug/Verbose level, and maybe one the first time an error comes in contacting a particular peer -- and when it succeeds again.

Which this refactor could help with, incidentally, but the real solution is log levels, not suppression, imo

**Comment by xiang90** [comment]
@barakmich The first error we met, we probably should log it anyway. The following ones, yea, we should not simply swallow them. When we have a verbose level logging, we should still log them. Correct?

**Comment by barakmich** [comment]
Right. Or, even better, instead of a timeout detection bit, log the first time the error shows up for this host/pair/message-type -- and then log a "all is well again" the next time it succeeds. Think of success as also an error log condition, which continues to be swallowed :)

**Comment by yichengq** [comment]
I agree that etcd should have log levels.
I think the suppression will still be used in debug/verbose level. Even if using a verbosal log, user is the one that reads the log, and it can be hard for him to read log including thousands of the same error lines.

**Comment by yichengq** [comment]
I can add the log line about "all starts to work".
I think timeout detection bit helps to say that the error is still happening and not fixed. I am open for better representation way.

**Comment by yichengq** [comment]
@barakmich "all starts to work" is added

**Comment by barakmich** [comment]
So I tried it, and was confused as to why it didn't (at first) appear to work. Because, running the goreman script and killing just one of the three members, I started to see this:
```
13:51:19 etcd2 | 2015/02/27 13:51:19 rafthttp: encountered error sending messages in stream msgapp from 6e3bd23ae5f1eae0 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:51:19 etcd1 | 2015/02/27 13:51:19 rafthttp: encountered error sending messages in stream msgapp from a8266ecf031671f3 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:51:19 etcd1 | 2015/02/27 13:51:19 rafthttp: encountered error sending messages in stream message from a8266ecf031671f3 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:51:24 etcd2 | 2015/02/27 13:51:24 rafthttp: encountered error sending messages in stream message from 6e3bd23ae5f1eae0 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:51:24 etcd2 | 2015/02/27 13:51:24 rafthttp: encountered error sending messages in pipeline from 6e3bd23ae5f1eae0 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:51:29 etcd2 | 2015/02/27 13:51:29 rafthttp: encountered error sending messages in stream msgapp from 6e3bd23ae5f1eae0 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:51:29 etcd1 | 2015/02/27 13:51:29 rafthttp: encountered error sending messages in stream msgapp from a8266ecf031671f3 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:51:29 etcd1 | 2015/02/27 13:51:29 rafthttp: encountered error sending messages in stream message from a8266ecf031671f3 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:51:34 etcd2 | 2015/02/27 13:51:34 rafthttp: encountered error sending messages in stream message from 6e3bd23ae5f1eae0 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
13:51:34 etcd2 | 2015/02/27 13:51:34 rafthttp: encountered error sending messages in pipeline from 6e3bd23ae5f1eae0 to 924e2e83e93f2560: dial tcp 127.0.0.1:7003: connection refused
```

You think I wouldn't see the same message from the same host, but every 10 seconds, there's two messages. 

Now, I get the reason why in the code: One's like a "heartbeat" and one's like a "connect". So really they're two errors at different locations. However, I cannot tell the difference; so either (a) the message itself should be deduplicated or (b) you should report what part (heartbeat, connect) is the one that's failing and being deduped. I rather like the former.

**Comment by barakmich** [comment]
NM that last comment. I see the subtle difference now. 

To that end "encountered error" can probably just be "error" -- but I'd still like to know when there was a failure in the stream. Was it trying to "connect", or "heartbeat", or what?

**Comment by yichengq** [comment]
@barakmich It prints out the behavior explicitly now:
```
failed to connect 924e2e83e93f2560 in stream message: dial tcp 127.0.0.1:7003: connection refused
```
Look better?

File: rafthttp/report.go [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+)
+
+var (
+	linkErrMapMu sync.Mutex
+	linkErrMap   = make(map[linkName]linkErr)
+)
+
+type linkName struct {
+	local, remote types.ID
+	worker        string
```
Comment (xiang90) [inline rafthttp/report.go]: can we not call this worker?

Comment (yichengq) [inline rafthttp/report.go]: maybe use `typename` or just `t`?

**Pull request closed without merging**